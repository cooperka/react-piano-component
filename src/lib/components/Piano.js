import React, { Fragment } from 'react';
import Instrument from './Instrument';
import isAccidentalNote from '../utils/isAccidentalNote';
import getNotesBetween from '../utils/getNotesBetween';
import getKeyboardShortcutForNote from '../utils/getKeyboardShortcutsForNote';

export default function Piano({
  startNote, endNote, keyboardMap, renderPianoKey, pianoKeyProps, renderAudio, ...instrumentProps
}) {
  const notes = getNotesBetween(startNote, endNote);
  const getNoteAtIndex = index => notes[index];

  return (
    <Instrument
      instrument={'acoustic_grand_piano'}
      keyboardMap={keyboardMap}
      renderInstrument={({ notesPlaying, onPlayNoteStart, onPlayNoteEnd }) =>
        notes.map((note, index) => (
          <Fragment key={note}>
            {renderPianoKey({
              note,
              index,
              isNoteAccidental: isAccidentalNote(note),
              isNotePlaying: notesPlaying.includes(note),
              startPlayingNote: () => onPlayNoteStart(note),
              stopPlayingNote: () => onPlayNoteEnd(note),
              keyboardShortcuts: getKeyboardShortcutForNote(keyboardMap, note),
              ...pianoKeyProps,
            })}
          </Fragment>
        ))
      }
      renderAudio={renderAudio}
      getNoteAtIndex={getNoteAtIndex}
      {...instrumentProps}
    />
  );
}

Piano.defaultProps = {
  keyboardMap: {},
};
