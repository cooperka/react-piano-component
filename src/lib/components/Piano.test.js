import React from 'react';
import { mount } from 'enzyme';
import Piano from './Piano';

function PianoKey({ note, startPlayingNote, stopPlayingNote }) {
  return <button data-note={note}
    onMouseDown={startPlayingNote}
    onMouseUp={stopPlayingNote}
  />;
}

function startPlayingPianoKey(wrapper, note) {
  wrapper.find(`button[data-note="${note}"]`).simulate('mousedown');
}

function stopPlayingPianoKey(wrapper, note) {
  wrapper.find(`button[data-note="${note}"]`).simulate('mouseup');
}

describe('Piano', () => {
  it('renders Instrument with the expected props', () => {
    const wrapper = mount(
      <Piano startNote={'C1'}
        endNote={'C2'}
        renderPianoKey={PianoKey}
      />
    );

    const instrumentProps = wrapper.childAt(0).props();
    expect(instrumentProps).toMatchSnapshot();
    expect(instrumentProps.getNoteAtIndex(0)).toMatchSnapshot();
  });

  it('calls renderPianoKey() with the expected props', () => {
    const renderPianoKey = jest.fn(PianoKey);
    mount(<Piano startNote={'C1'}
      endNote={'C2'}
      renderPianoKey={renderPianoKey}
      pianoKeyProps={{ test: true }}
    />);

    expect(renderPianoKey.mock.calls).toMatchSnapshot();
  });

  it('calls renderAudio() with the expected props', () => {
    const renderAudio = jest.fn(() => null);
    const wrapper = mount(
      <Piano startNote={'C1'}
        endNote={'C2'}
        renderPianoKey={PianoKey}
        renderAudio={renderAudio}
      />
    );

    startPlayingPianoKey(wrapper, 'C1');
    startPlayingPianoKey(wrapper, 'C2');
    stopPlayingPianoKey(wrapper, 'C1');

    expect(renderAudio.mock.calls).toMatchSnapshot();
  });
});
