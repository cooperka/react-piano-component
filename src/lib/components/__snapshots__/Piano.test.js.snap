// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Piano calls renderAudio() with the expected props 1`] = `
Array [
  Array [
    Object {
      "notes": Array [],
    },
    Object {},
  ],
  Array [
    Object {
      "notes": Array [
        "C1",
      ],
    },
    Object {},
  ],
  Array [
    Object {
      "notes": Array [
        "C1",
        "C2",
      ],
    },
    Object {},
  ],
  Array [
    Object {
      "notes": Array [
        "C2",
      ],
    },
    Object {},
  ],
]
`;

exports[`Piano calls renderPianoKey() with the expected props 1`] = `
Array [
  Array [
    Object {
      "index": 0,
      "isNoteAccidental": false,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "C1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 1,
      "isNoteAccidental": true,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "C#1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 2,
      "isNoteAccidental": false,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "D1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 3,
      "isNoteAccidental": true,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "D#1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 4,
      "isNoteAccidental": false,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "E1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 5,
      "isNoteAccidental": false,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "F1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 6,
      "isNoteAccidental": true,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "F#1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 7,
      "isNoteAccidental": false,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "G1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 8,
      "isNoteAccidental": true,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "G#1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 9,
      "isNoteAccidental": false,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "A1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 10,
      "isNoteAccidental": true,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "A#1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 11,
      "isNoteAccidental": false,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "B1",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
  Array [
    Object {
      "index": 12,
      "isNoteAccidental": false,
      "isNotePlaying": false,
      "keyboardShortcuts": Array [],
      "note": "C2",
      "startPlayingNote": [Function],
      "stopPlayingNote": [Function],
      "test": true,
    },
  ],
]
`;

exports[`Piano renders Instrument with the expected props 1`] = `
Object {
  "getNoteAtIndex": [Function],
  "instrument": "acoustic_grand_piano",
  "keyboardMap": Object {},
  "renderAudio": undefined,
  "renderInstrument": [Function],
}
`;

exports[`Piano renders Instrument with the expected props 2`] = `"C1"`;
